### PR TITLE
CI: Build a xDSL-embedding Pyodide distribution.

### PR DESCRIPTION
**Pull request by PapyChacal**
This PR now builds a custom Pyodide distribution, embedding xDSL and dependencies, so we can just `import xdsl` in JupyterLite, as desired. It should also improve startup time.
This can be previewed at [papychacal.github.io/xdsl](https://papychacal.github.io/xdsl)
Note you might need to clear your browser's cache for the page, especially if you get some "hash mismatch" error on importing xDSL.

The Pyodide build is cached, so it should take less than 5 minutes once cached.

### CODE REVIEW

File: .github/workflows/jupyterlite.yml [hunk 1]
```diff
@@ -11,68 +11,69 @@ jobs:
...
+          python -m pip install -r requirements.txt
+          sudo apt update && sudo apt install f2c
+          
+          rm packages/xdsl packages/frozenlist -rf
+          pyodide skeleton pypi xdsl
+          pyodide skeleton pypi frozenlist
+          
+          python -c 'import yaml; f = open("packages/xdsl/meta.yaml"); y = yaml.safe_load(f); f.close(); y["requirements"] = {"run" : ["frozenlist"]}; y["source"] = {"url" : "'$XDSL_SOURCE_URL'", "sha256" : "'$XDSL_SHA256'"}; f = open("packages/xdsl/meta.yaml", "w"); yaml.dump(y, f)'
+          
+          PYODIDE_PACKAGES="frozenlist,coverage,xdsl" make
```
Comment (webmiche) [inline .github/workflows/jupyterlite.yml:60]: I don't understand everything written here and why it is necessary, but isn't there a github action for [pyodide](https://github.com/jobovy/pyodide-buildpackage-action)? Why not use that?

**Review by webmiche — COMMENTED** [review]
Nice, this is certainly a good step in the right direction!

File: .github/workflows/jupyterlite.yml [hunk 2]
```diff
@@ -11,68 +11,69 @@ jobs:
...
+          git fetch --all
+          git checkout 0.22.0a3
+          python -m pip install -r requirements.txt
+          sudo apt update && sudo apt install f2c
+          
+          rm packages/xdsl packages/frozenlist -rf
+          pyodide skeleton pypi xdsl
+          pyodide skeleton pypi frozenlist
+          
+          python -c 'import yaml; f = open("packages/xdsl/meta.yaml"); y = yaml.safe_load(f); f.close(); y["requirements"] = {"run" : ["frozenlist"]}; y["source"] = {"url" : "'$XDSL_SOURCE_URL'", "sha256" : "'$XDSL_SHA256'"}; f = open("packages/xdsl/meta.yaml", "w"); yaml.dump(y, f)'
```
Comment (webmiche) [inline .github/workflows/jupyterlite.yml:58]: Why is there inline python code here?

File: .github/workflows/jupyterlite.yml (hunk 1 above)
Comment (PapyChacal) [reply .github/workflows/jupyterlite.yml:60]: The pyodide action does things similarly to how things worked before this PR.
Thing is, JupyterLite doesn't allow us to properly configure the code loading Pyodide on their side, so using that would require some hacking and rebuilding in JupyterLite.
Rebuilding Pyodide with xDSL in-tree is currently the cleanest solution (Even recommended [in their doc](https://jupyterlite.readthedocs.io/en/latest/howto/python/packages.html#bundling-additional-packages-by-default))

File: .github/workflows/jupyterlite.yml (hunk 2 above)
Comment (PapyChacal) [reply .github/workflows/jupyterlite.yml:58]: This is to update the meta.yml file used by Pyodide to build and use xDSL, so it points to the local xDSL sources and states the runtime dependency to frozenlist.

I'll take that logic out in a separate python script as discussed in person.

**Review by webmiche — APPROVED** [review]

**Review by georgebisbas — COMMENTED** [review]
One minor proposal, than another review and we could merge I guess

File: .github/workflows/update_xdsl_pyodide_build.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import sys
+import os
+import hashlib
+
+meta_yaml_path = sys.argv[1]
+xdsl_directory = sys.argv[2]
+
+# Parse the auto-generated one
+with open(meta_yaml_path) as f:
+    y = yaml.safe_load(f)
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply .github/workflows/update_xdsl_pyodide_build.py:16]: Sure!

**Review by math-fehr — APPROVED** [review]
Nice! Looks good to me!

**Pull request merged**